django/contrib/contenttypes/views.py

```diff
@@ -43,27 +43,37 @@ def shortcut(request, content_type_id, object_id):
 
     # Otherwise, we need to introspect the object's relationships for a
     # relation to the Site object
-    object_domain = None
+    try:
+        object_domain = get_current_site(request).domain
+    except ObjectDoesNotExist:
+        object_domain = None
 
     if apps.is_installed('django.contrib.sites'):
         Site = apps.get_model('sites.Site')
+        check_many_to_one = True
 
         opts = obj._meta
 
         # First, look for a many-to-many relationship to Site.
         for field in opts.many_to_many:
             if field.remote_field.model is Site:
-                try:
-                    # Caveat: In the case of multiple related Sites, this just
-                    # selects the *first* one, which is arbitrary.
-                    object_domain = getattr(obj, field.name).all()[0].domain
-                except IndexError:
-                    pass
-                if object_domain is not None:
+                site_qs = getattr(obj, field.name).all()
+
+                if object_domain and site_qs.filter(domain=object_domain).exists():
+                    # The current site's domain matches a site attached to the object.
+                    check_many_to_one = False
                     break
 
+                # Caveat: In the case of multiple related Sites, this just
+                # selects the *first* one, which is arbitrary.
+                site = site_qs.first()
+
+                if site:
+                    object_domain = site.domain
+                    check_many_to_one = False
+
         # Next, look for a many-to-one relationship to Site.
-        if object_domain is None:
+        if check_many_to_one:
```

Comment:
I think this could be replaced with for/else and adding `break` where `check_many_to_one = False` is.